### PR TITLE
[ACR-1579] Add machine-learning-training docs

### PR DESCRIPTION
**Pull request by zhoonit**
**Change Description**
- Add machine-learning-training docs
- Update overview.md accordingly

**API Changes**
- ACR: [ACR-1579](https://code.sec.samsung.net/jira/browse/ACR-1579)
Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

**Revisions**
V2. Fix typos, expressions, grammar (5be9a1e -> b4465cf)
V3. Add documentation about available properties (b4465cf ->  ba06a1b)
V4. Change writing styles and tones to reflect reviews ( ba06a1b ) -> ( bc6ac55 )
V5. Fix note style and expressions ( bc6ac55 ) -> ( 1614d20 ) 
V6. Update feature key to the system.md ( 1614d20 ) -> ( 13448bb )

### CODE REVIEW

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 1]
```diff
@@ -0,0 +1,402 @@
+# Machine Learning Training
+
+Machine Learning Training allows to construct, control and train a machine learning model in Tizen devices.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:3]: Machine Learning Training `API/system/....`?

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+Please note that:
+
+* This feature remains experimental for Tizen 6.0 and is not full-fledged yet for every needs.
+* Every example code is not fully handling error for brevity. Error should be handled more extensively then example code written here.
+
+## Prerequisites
+
+To enable your application to use the machine learning training functionality:
+
+1. To use the functions and data types of the Machine Learning Inference API, include the `<nntrainer.h>` header file in your application:
```
Comment (jaeyun-jung) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:14]: Machine Learning Inference API
change to train

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+If you have file that describe model, the file can be used to construct initially with `ml_train_model_construct_with_file`.
+Even if you constructed model from a file, you can switch/modify/add each component with set of given apis until you compile with `ml_train_model_compile`.
+
+### Construct Model from the Description File
+
+As of now, only `*.ini` is supported to construct model from a file.
+
+#### Create Model from ini Format
+
+There are special sections refering to `model` and `dataset` respectively [Model] and [Dataset]
```
Comment (jaeyun-jung) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:139]: maybe need to add links [Model] and [Dataset]

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 2 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:14]: Done

**Comment by zhoonit** [comment]
V2. Fix typos, expressions, grammar

**Comment by zhoonit** [comment]
v3. Add documentation about available properties (b4465cf ->  ba06a1b)

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+&#xfeff;                            | decay_steps        | (float)       | Decay steps
+&#xfeff;                            | decay_rate         | (float)       | Decay rate
+&#xfeff;                            | beta1              | (float)       | beta1 coefficient for adam
+&#xfeff;                            | beta2              | (float)       | beta2 coefficient for adam
+&#xfeff;                            | epsilon            | (float)       | epsilon coefficient for adam
+
+### Dataset
+
+Dataset is in charge of feeding data into the model.
+It can either be created from a callback function or created from a file.
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:168]: Please change to:

The dataset can either be.....

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+
+// configure optimizer
+ml_train_optimizer_set_property(optimizer, "learning_rate=0.001", NULL);
+
+// after configuring optimizer to model
+// you do not need to destroy optimizer since ownership is transferred to the model.
+ml_train_model_set_optimizer(model, optimizer);
+```
+
+Available properties for each optimizer type are as follows:
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:151]: Suggestion:

Following are the available properties for each optimizer type:

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+Type = fully_connected
+Unit = 2    # Output Layer Dimension ( = Weight Width )
+bias_initializer = zeros
+weight_initializer = xavier_uniform
+Activation = sigmoid  # activation : sigmoid, softmax
+Weight_Decay = l2norm
+weight_Decay_Lambda = 0.005
+```
+
+Some restrictions must be kept:
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:246]: The following restrictions must be adhered to:

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+
+ * Model file must have a `[Model]` section.
+ * Model file must have at least one layer.
+ * Valid keys must have valid properties (while invalid keys in each section will be ignored.)
+
+Please note that all paths are relative to the current working directory unless stating an absolute path.
+It is recommended to set `save_path` and `Dataset` from the code over describing inside the model file.
+
+Following example constructs model from INI file
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:255]: The following example constructs model from an INI file:

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+
+## Configure the Model
+
+After constructing a model, the model can be configured. This section provides example code to reproduce model description from [Create Model from INI Formatted File](#create-model-from-ini-formatted-file)
+
+Except that:
+ * Relative path is changed to dynamic app resource and data path.
+ * Model related properties (which can be set at compile / run phase)
+ * Demonstration about `ml_train_dataset_create_with_generator` which can't be covered in the description file.
+
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:283]: Please rephrase to use the Tizen writing style and tone.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+```c
+int status = ML_ERROR_NONE;
+
+status = ml_train_model_run(model, "epochs=20", "save_path=model.bin", NULL);)
+```
+
+## Destroy the Model
+
+After training, the model should be destroyed with `ml_train_model_destroy`.
+Note that `ml_train_model_add_layer`, `ml_train_set_optimizer`, `ml_train_set_dataset` transferrers ownership to `model`.
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:453]: Please change the sentence structure and use Note tag

**Review by surbhig12 — COMMENTED** [review]
Minor comment. Please update so that the PR can be approved.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+  ```xml
+  <privileges>
+    <!-- For accessing media storage -->
+    <privilege>http://tizen.org/privilege/mediastorage</privilege>
+    <!-- For accessing external storage -->
+    <privilege>http://tizen.org/privilege/externalstorage</privilege>
+  </privileges>
+  ```
+
+## Building Blocks
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:46]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+
+There are 4 major components in Machine Learning Training API.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:48]: Changed as per suggestion.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+
+### Model
+
+![model](./media/model.png)
+
+Model is a wrapper component that has the topology of layers, optimizers, datasets.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:54]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+The model performs training and saves the updated parameters that can later be used for inference.
+Please note that only a sequential neural network is supported for the time being.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:56]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:57]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 15]
```diff
@@ -0,0 +1,460 @@
...
+// destroy after use
+ml_train_model_destroy(model);
+```
+
+A number of properties can be set at `ml_train_model_compile` and `ml_train_model_run` phase:
+
+Function | Key | Value | Description
+-------- | --- | ----- | -----------
+`ml_train_model_compile` | loss      | (categorical)| Loss function to be set
+&#xfeff;                 |           | cross        | cross entropy loss
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:79]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 16]
```diff
@@ -0,0 +1,460 @@
...
+// after set dataset to model,
+// you do not need to destroy dataset since ownership is transferred to the model.
+ml_train_model_set_dataset(model, dataset);
+```
+
+## Construct a Model
+
+A model can be constructed with `ml_train_model_construct`.
+If you have a file that describes the model, the file can be used to construct initially with `ml_train_model_construct_with_file`.
+Even if the model is constructed from a file, Switching/modifying/setting a component is possible until compiling with `ml_train_model_compile`.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:188]: Changed accordingly

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 6 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:246]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 8 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:283]: I skimmed through https://github.com/Samsung/tizen-docs/tree/master/styleguide but, I am not sure what defines `Tizen writing style and tone` . I rephrased a bit though. Could you elaborate more on this topic if modification isn't enought?

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 17]
```diff
@@ -0,0 +1,460 @@
...
+First, an empty model needs to be created.
+
+```c
+ml_train_model_h model;
+ml_train_model_construct(&model);
+```
+
+### Add a Layer
+
+`ml_train_model_add_layer` appends a layer to the end of the graph in the model.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:293]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 18]
```diff
@@ -0,0 +1,460 @@
...
+if(status != ML_ERROR_NONE) {
+  // handle error
+}
+
+status = ml_train_dataset_property(dataset, label_prop,
+                                   "buffer_size=9", NULL);
+status = ml_train_model_set_dataset(model, dataset);
+```
+
+Note that `buffer_size` can be set for dataset.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:394]: Could you add reasoning for this? 

In the style guide, [!NOTE] is for information the user should notice even if skimming. This is not notable information rather an explanation about the code above.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 9 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:453]: Done

File: docs/application/native/guides/machine-learning/overview.md [hunk 19]
```diff
@@ -7,7 +7,11 @@ You can use the following machine learning features in your native applications:
 
 - [Machine Learning Inference](machine-learning-inference.md)
 
-  You can create data pipelines that contain the machine learning model filters. The pipeline can be assembled and controlled with various elements of [GStreamer](https://gstreamer.freedesktop.org/). 
+  You can create data pipelines that contain the machine learning model filters. The pipeline can be assembled and controlled with various elements of [GStreamer](https://gstreamer.freedesktop.org/).
+
+- [Machine Learning Training](machine-learning-training.md)
+
+  You can create trainable models on-the-fly inside the code or with a set of format that describes model. This models can be trained on the device locally when needed.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/overview.md:14]: Changed as 

  You can construct, control, and train a machine learning model on-the-fly. The training can be done using the code or with a set of format that describes the model. These models can be trained on the device locally when needed.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 20]
```diff
@@ -0,0 +1,460 @@
...
+### Set a Dataset
+
+There are two ways to create a dataset. One is from a file, and another one is from a callback.
+In either fashion, you need to provide streams of tensor data and one-hot-encoded labels.
+
+#### Set a Dataset from a File
+
+To create a `dataset` from a file, a training set and a label set must be provided. A validation set and a test set can be provided if needed.
+
+A label set is a file that has actual labels each in a line. Please refer to the example below.
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:349]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 21]
```diff
@@ -0,0 +1,460 @@
...
+3. In case of saving/loading model files from the outside of the application's own resources, the application has to request permission by adding following privileges to the `tizen-manifest.xml` file:
+
+  ```xml
+  <privileges>
+    <!-- For accessing media storage -->
+    <privilege>http://tizen.org/privilege/mediastorage</privilege>
+    <!-- For accessing external storage -->
+    <privilege>http://tizen.org/privilege/externalstorage</privilege>
+  </privileges>
+  ```
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:44]: Done

**Review by zhoonit — COMMENTED** [review]
Changed accordingly, PTAL

**Comment by zhoonit** [comment]
v3. Change writing styles and tones to reflect reviews ( ba06a1b ) -> ( bc6ac55 )

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 22]
```diff
@@ -0,0 +1,468 @@
...
+// after setting a dataset to model,
+// you do not need to destroy dataset since ownership is transferred to the model.
+ml_train_model_set_dataset(model, dataset);
+```
+
+## Construct a model
+
+A model can be constructed with `ml_train_model_construct`.
+If you have a file that describes the model, the file can be used to construct initially with `ml_train_model_construct_with_file`.
+Even if the model is constructed from a file, Switching, modifying or setting a component is possible until compiling with `ml_train_model_compile`.
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:194]:
```suggestion
Even if the model is constructed from a file, switching, modifying or setting a component is possible only until you compile with `ml_train_model_compile`.
```

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 23]
```diff
@@ -0,0 +1,468 @@
...
+weight_regularizer_constant = 0.005
+```
+
+The following restrictions must be adhered to:
+
+ * Model file must have a `[Model]` section.
+ * Model file must have at least one layer.
+ * Valid keys must have valid properties (while invalid keys in each section will be ignored.)
+
+[!NOTE]
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:258]: Global: Please add the right angular bracket before the note and the sentence:

> [!NOTE]
> All paths are relative to the current working directory unless stating an absolute path. It is recommended to set `save_path` and `Dataset` from the code over describing inside the model file.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 24]
```diff
@@ -0,0 +1,468 @@
...
+
+### Construct a model on code
+
+An empty model can be constructed with `ml_train_model_construct`.
+
+## Configure the model
+
+After constructing a model, the model can be configured.
+
+[!NOTE]
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:286]: Please fix the usage of Note style in the entire file.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (surbhig12) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:400]: Thank you for the clarification. In that case, please remove the word "Note that".

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 25]
```diff
@@ -0,0 +1,468 @@
...
+
+```c
+int status = ML_ERROR_NONE;
+
+status = ml_train_model_run(model, "epochs=20", "save_path=model.bin", NULL);)
+```
+
+## Destroy the model
+
+After training, the model should be destroyed with `ml_train_model_destroy`.
```
Comment (surbhig12) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:458]:
```suggestion
After training, the model must be destroyed with `ml_train_model_destroy`.
```

**Review by surbhig12 — COMMENTED** [review]
Reviewed updated comments.

**Comment by zhoonit** [comment]
Changed as per request. Please take another look @surbhig12

**Comment by zhoonit** [comment]
V5. Fix note style and expressions ( bc6ac55 ) -> ( 1614d20 )

**Comment by zhoonit** [comment]
V6. Update feature key to the system.md ( 1614d20 ) -> ( 13448bb )

**Comment by zhoonit** [comment]
Closing this PR as target branch should be tizen_6.0_prepare.
